### PR TITLE
Fix aslr for ARMv7 smp

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This PR touches platforms maintained by:
@etienne-lms, @MrVan, @cneveux, @sdininno, @clementfaure, @pangupta, @sahilnxp, @nodeax, , @grandpaul, @Amit-Radur, @OP-TEE/plat-zynq7k

### CODE REVIEW

File: core/arch/arm/plat-stm/main.c [hunk 1]
```diff
@@ -127,18 +127,19 @@ void arm_cl2_config(vaddr_t pl310)
...
 
-void plat_cpu_reset_late(void)
+void plat_primary_init_early(void)
 {
 	int i;
 
 	assert(!cpu_mmu_enabled());
 
+	/*
+	 * TODO: find out if this can be done in plat_cpu_reset_early()
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/main.c]: @etienne-lms

**Comment by jenswi-linaro** [comment]
The travis error is a false positive in checkpatch.

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 2]
```diff
@@ -264,12 +264,18 @@ KEEP_INIT _start
 		cmp	r0, #CFG_TEE_CORE_NB_CORE
 		/* Unsupported CPU, park it before it breaks something */
 		bge	unhandled_cpu
-		ldr	r1, =stack_tmp_stride
-		ldr	r1, [r1]
+
+		adr	r3, stack_tmp_stride_rel
+		ldr	r1, [r3]
+		ldr	r1, [r1, r3]
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: I don't catch this trick. Shouldn't it be `add r1, r1, r3` at lien 270? (and equiv at line 275)

Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_entry_a32.S]: OK, I'll add a comment. At the address of `stack_tmp_stride_rel`, is the offset from that address to `__stack_tmp_stride` stored. `add r1, r1, r3` would compute the address of `__stack_tmp_stride`, while `ldr r1, [r1, r3]` does a load from that address.

Comment (etienne-lms) [reply core/arch/arm/kernel/generic_entry_a32.S]: oh, right!

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 3]
```diff
@@ -727,12 +741,23 @@ KEEP_PAGER cpu_on_handler
...
 
 	wait_primary
 
 	set_sp
+#ifdef CFG_CORE_ASLR
+	/*
+	 * stack_tmp_export from which is used as base when initializing sp
+	 * has been relocated to the new offset. Since mmu isn't enabled on
+	 * this cpu yet we need to restore the physical corresponding
+	 * physical address.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_entry_a32.S]: Thanks, I guess I was exhausted from getting this to work.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jenswi-linaro** [comment]
Squashed in the fixups and applied tags.

**Comment by jforissier** [comment]
Ignoring the two checkpatch warnings -- we probably don't want to fix the one about `KEEP_PAGER()` anyways.

**Pull request merged**